### DISCUSSION TITLE
如何使用包管理器下载已有的预编译包

### DISCUSSION THREAD

**Discussion by BloodingKnight** [xmake-io/xmake#5427] [Q&A]
我有一些预编译包，手动下载到本地后，可以通过添加package的方式正常将他们加入到构件中，我该如何修改我的package 配置，让xmake能够自动帮我下载呢？
```lua
package("hcl")
    set_sourcedir(path.join(os.scriptdir(), "3rd_party/xxx"))
    on_load(function (package)
        package:set("installdir", os.scriptdir())
        package:add("links", "xxx")
    end)
    on_fetch(function (package)
        local result = {}
        result.links = "xxx"
        result.linkdirs = path.join(package:installdir("lib"), package:arch())
        result.includedirs = package:installdir("include")
        return result
    end)
package_end()
```

我尝试直接用 `add_urls` 与 `add_versions` 但似乎安装后~/.xmake/packages/x/xxx 目录是空的

**Comment by BloodingKnight** [comment]
有一个额外的需求就是，如果远程仓库中没有已经编译好的包，是否可以访问某个指定的url呢？

**Comment by waruqi** [comment]
> 我该如何修改我的package 配置，让xmake能够自动帮我下载呢？

走自动下载，看下 xmake-repo 里面的配置不就好了，里面全是的。

**Reply by BloodingKnight** [reply]
我是按照gtest的xmake.lua写的，不过这个似乎是下载源码？我需要的是下载预编译好的包

**Reply by waruqi** [reply]
有区别么，url 里面 tar 包文件，不是你自己决定的么，放源码 还是 放预编译文件 不都是一样装么

**Reply by BloodingKnight** [reply]
我不知道这两个有没有区别啊，现在情况就是我按照上面的写了之后，~/.xmake/package 对应文件夹下是空的，所以我才来询问的呀

**Reply by BloodingKnight** [reply]
或者有没有文档描述一下xmake 在遇到package时，on_install/ on_load / on_fetch 具体做了什么事情，整个下载-解压-移动安装-引入是怎么各流程？

**Reply by waruqi** [reply]
https://xmake.io/#/zh-cn/package/remote_package?id=%e5%ae%89%e8%a3%85%e4%ba%8c%e8%bf%9b%e5%88%b6%e5%8c%85

**Reply by BloodingKnight** [reply]
额，找到问题了，似乎是我的下载链接地址是 http，不支持https，而xmake似乎会尝试以 https 协议来下载，请问这个是 xmake 的设计吗？是否有地方修改呢？

**Reply by BloodingKnight** [reply]
还有第二个问题，version字符串不能带下划线吗？

**Reply by star-hengxing** [reply]
https://github.com/xmake-io/xmake-repo/blob/f8bae8b14d98c2d9f7c7ca367e76c25c9360cd93/packages/l/libyuv/xmake.lua#L9-L17

**Reply by waruqi** [reply]
> 额，找到问题了，似乎是我的下载链接地址是 http，不支持https，而xmake似乎会尝试以 https 协议来下载，请问这个是 xmake 的设计吗？是否有地方修改呢？

这样问 curl/wget了， xmake 仅仅只是简单的调用命令下载，不会区分是否 https

**Reply by BloodingKnight** [reply]
已经解决了，主要遇到了下面几个问题：

1. version 在解析时仅支持 "." 与 "-" 作为特殊符号，如果加入了 "_" 后，version就会被置为空，虽然 version_str 依旧可以正常访问，但无法在 add_urls 中使用
2. 如果想要自动下载，就不能定义 `on_fetch` 
3. 下载预编译包时，version的第二个字段必须是shasum

**Reply by waruqi** [reply]
> 如果想要自动下载，就不能定义 on_fetch

不对，默认优先 fetch 系统包，只有 on_fetch 返回空，才会自动下，而不是定义。

xmake-repo 很多包都是有 on_fetch 定义的。。另外，也可以 system = false 禁用系统包来实现

**Reply by BloodingKnight** [reply]
> 只有 on_fetch 返回空

这里说的是返回了空列表，还是目录中没有对应的文件呢？

**Reply by waruqi** [reply]
返回 nil
